worker/raft/rafttransport/handler.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package rafttransport
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+
+	"github.com/juju/errors"
+	"github.com/juju/juju/apiserver/apiserverhttp"
+)
+
+// Handler is an http.Handler suitable for serving an endpoint that
+// upgrades to raft transport connections.
+type Handler struct {
+	connections chan<- net.Conn
+	abort       <-chan struct{}
+}
+
+// NewHandler returns a new Handler that sends connections to the
+// given connections channel, and stops accepting connections after
+// the abort channel is closed.
+func NewHandler(
+	connections chan<- net.Conn,
+	abort <-chan struct{},
+) *Handler {
+	return &Handler{
+		connections: connections,
+		abort:       abort,
+	}
+}
+
+// ServeHTTP is part of the http.Handler interface.
+//
+// ServeHTTP checks for "raft" upgrade requests, and hijacks
+// those connections for use as a raw connection for raft
+// communications.
+func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
```

Comment:
I think this is the code which is lifted from https://github.com/CanonicalLtd/raft-http?
Could we indicate that in a comment?

Reply:
done